include/pybind11/pybind11.h

```diff
@@ -650,8 +650,15 @@ class cpp_function : public function {
                         result = PYBIND11_TRY_NEXT_OVERLOAD;
                     }
 
-                    if (result.ptr() != PYBIND11_TRY_NEXT_OVERLOAD)
+                    if (result.ptr() != PYBIND11_TRY_NEXT_OVERLOAD) {
+                        if (!result) {
+                            // The error reporting logic below expects 'it' to be valid, as it would be
+                            // if we'd encountered this failure in the first-pass loop.
+                            for (it = overloads; it != &call.func; it = it->next)
```

Comment:
ISTM this could be simplified down to just:
```C++
if (!result)
    it = &call.func;
```
(unless I'm forgetting something about this code).

Reply:
D'oh, right you are - I was getting confused about function_call vs function_record.